File: python/hot100/74.数组中第K个最大元素.py
class Solution:
    # 启用三路快排分区
    def partition(self, nums: List[int], left: int, right: int) -> tuple[int, int]:
        # 1. 随机选择基准，并交换到 left (修正语法)
        random_index = random.randint(left, right)
        nums[left], nums[random_index] = nums[random_index], nums[left]
        
        pivot = nums[left] # 2. 正确获取基准值
        
        lt = left      # nums[left..lt] < pivot 的区域末尾
        gt = right     # nums[gt..right] > pivot 的区域开头
        i = left + 1   # nums[lt+1..i-1] = pivot 的区域
        
        while i <= gt:
            if nums[i] < pivot:
                # 将 < pivot 的元素换到 = pivot 区域的开头
                nums[i], nums[lt + 1] = nums[lt + 1], nums[i]
                lt += 1
                i += 1
            elif nums[i] > pivot:
                # 将 > pivot 的元素换到 gt 位置
                nums[i], nums[gt] = nums[gt], nums[i]
                gt -= 1 
                # i 不动，因为新换回来的 nums[i] 还需要检查
            else:
                # = pivot 的元素，继续向右扫描
                i += 1
                
        # 将基准值换到正确位置：换到小于区域的末尾
        # 此时 lt 是小于区域的最后一个元素索引，交换后 nums[lt] = pivot
        nums[left], nums[lt] = nums[lt], nums[left]
        
        # 返回等于 pivot 元素的范围 [lt, gt]
        return lt, gt

    def findKthLargest(self, nums: List[int], k: int) -> int:
        # 必须在调用前初始化随机数生成器，虽然 random.randint() 通常不需要，但保持习惯是好的。
        # random.seed() # 如果需要可重现的结果，可以设置种子
        
        n = len(nums)
        target = n - k # 第 k 大转换为第 n-k 小 (索引)
        left, right = 0, n - 1
        
        while left <= right:
            lt, gt = self.partition(nums, left, right)
            
            if target >= lt and target <= gt:
                # 目标索引在等于 pivot 的区域内，找到了！
                return nums[target]
            elif target < lt:
                # 目标在小于 pivot 的区域内
                right = lt - 1
            else: # target > gt
                # 目标在大于 pivot 的区域内
                left = gt + 1
                
        return -1 # 理论上不会发生

File: cpp/hot100/74.数组中第K个最大元素.cpp
class Solution {
pair<int, int> partition(vector<int>& nums, int left, int right) {
    swap(nums[left + rand() % (right - left + 1)], nums[left]);
    int pivot = nums[left];
    int lt = left;      // nums[left..lt] < pivot
    int gt = right;     // nums[gt..right] > pivot
    int i = left + 1;   // nums[lt+1..i-1] = pivot
    while (i <= gt) {
        if (nums[i] < pivot) {
            // 小于基准值，与 lt+1 交换，lt 和 i 同步右移
            swap(nums[i], nums[lt + 1]);
            lt++;
            i++;
        } else if (nums[i] > pivot) {
            // 大于基准值，与 gt 交换，gt 左移，i 不动 (因为换回来的 nums[i] 还没检查)
            swap(nums[i], nums[gt]);
            gt--;
        } else {
            // 等于基准值，i 右移
            i++;
        }
    }
    
    // 将基准值换到正确位置
    swap(nums[left], nums[lt]);
    
    return {lt, gt}; // 返回等于 pivot 的元素的范围
}

public:
    int findKthLargest(vector<int>& nums, int k) {
        srand(time(NULL));
        int n = nums.size();
        int target = n - k; // 目标索引 (第 k 大转换为第 n-k 小)
        int left = 0, right = n - 1;

        while (left <= right) {
            pair<int, int> range = partition(nums, left, right);
            int start_idx = range.first;  // = pivot 区域的起始
            int end_idx = range.second;   // = pivot 区域的结束
            
            if (target >= start_idx && target <= end_idx) {
                // 目标索引在 = pivot 的区域内，找到了！
                return nums[target];
            } else if (target < start_idx) {
                // 目标在 < pivot 的区域内
                right = start_idx - 1;
            } else { // target > end_idx
                // 目标在 > pivot 的区域内
                left = end_idx + 1;
            }
        }
        return -1; // 不可能发生
    }
};